### PR TITLE
[SPARK-49738][SQL] Endswith bug fix

### PR DESCRIPTION
**Pull request by viktorluc-db**
### What changes were proposed in this pull request?

Bugfix in "endswith" string predicate. Also fixed the same type of the bug in `CollationAwareUTF8String.java` in method `lowercaseMatchLengthFrom`.

### Why are the changes needed?

Expression `select endswith('İo' collate utf8_lcase, 'İo' collate utf8_lcase)`
returns `false` (incorrect), but should return `true` (correct).

### Does this PR introduce _any_ user-facing change?

Yes, `endswith` function now returns proper results in the edge cases included above.

### How was this patch tested?

Added tests in CollationSupportSuite.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by viktorluc-db** [comment]
@MaxGekk could you review please?

**Comment by MaxGekk** [comment]
cc @uros-db

File: common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java [hunk 1]
```diff
@@ -880,6 +880,8 @@ public void testEndsWith() throws SparkException {
     assertEndsWith("the İo", "Io", "UTF8_LCASE", false);
     assertEndsWith("the İo", "i\u0307o", "UTF8_LCASE", true);
     assertEndsWith("the İo", "İo", "UTF8_LCASE", true);
+    assertEndsWith("İo", "İo", "UTF8_LCASE", true);
+    assertEndsWith("İo", "i̇o", "UTF8_LCASE", true);
```
Comment (uros-db) [inline common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java:884]: do we need this for `assertStartsWith` as well? given that you introduced a fix to `lowercaseMatchLengthFrom` as well

Comment (MaxGekk) [reply common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java:884]: +1

Comment (viktorluc-db) [reply common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java:886]: added

**Review by uros-db — APPROVED** [review]
tnx for this fix, @MaxGekk please review asap

**Review by MaxGekk — APPROVED** [review]

**Review by MaxGekk — CHANGES_REQUESTED** [review]
@viktorluc-db Please, update the golden files:
```
[info] - collations.sql *** FAILED *** (10 seconds, 643 milliseconds)
[info]   collations.sql
[info]   Expected "false
[info]   false
[info]   false
[info]   [false
[info]   fals]e
[info]   true
[info]   true
[info]   true
[info]   tru...", but got "false
[info]   false
[info]   false
[info]   [true
[info]   tru]e
[info]   true
[info]   true
[info]   true
[info]   tru..." Result did not match for query #164
```

**Comment by viktorluc-db** [comment]
@MaxGekk regenerated golden files

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @viktorluc-db and @uros-db for review.

**Comment by cloud-fan** [comment]
Let's make the PR title more precise next time, at least it should mention it's a bug fix for Endswith with string collation.

**Comment by gatorsmile** [comment]
@MaxGekk @viktorluc-db @uros-db , the results are different after the fix. We should also update the section: "Does this PR introduce any user-facing change?"

**Pull request closed without merging**